File: benchmark/cases/oop_dispatch/oop_dispatch.py
class Base:
    def compute(self):
        return 0

class TypeA(Base):
    def compute(self):
        return 1

class TypeB(Base):
    def compute(self):
        return 2

class TypeC(Base):
    def compute(self):
        return 3

class TypeD(Base):
    def compute(self):
        return 4

classes = [TypeA, TypeB, TypeC, TypeD]
s = 0
for i in range(100000):
    obj = classes[i % 4]()
    s += obj.compute()
print(s)

File: benchmark/cases/oop_dispatch/oop_dispatch.rs
trait Computable {
    fn compute(&self) -> i64;
}

struct TypeA;
struct TypeB;
struct TypeC;
struct TypeD;

impl Computable for TypeA { fn compute(&self) -> i64 { 1 } }
impl Computable for TypeB { fn compute(&self) -> i64 { 2 } }
impl Computable for TypeC { fn compute(&self) -> i64 { 3 } }
impl Computable for TypeD { fn compute(&self) -> i64 { 4 } }

fn main() {
    let mut objects: Vec<Box<dyn Computable>> = Vec::with_capacity(100_000);
    for i in 0..100_000i64 {
        let obj: Box<dyn Computable> = match i % 4 {
            0 => Box::new(TypeA),
            1 => Box::new(TypeB),
            2 => Box::new(TypeC),
            3 => Box::new(TypeD),
            _ => unreachable!(),
        };
        objects.push(obj);
    }
    let sum: i64 = objects.iter().map(|o| o.compute()).sum();
    println!("{}", sum);
}
